File: agent_wiki/workspace.py
from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass
from datetime import datetime, timezone
from fnmatch import fnmatch
from pathlib import Path
from typing import Iterable

from .config import DEFAULT_WORKSPACE_WIKI_DIR, AgentWikiConfig, WorkspaceScanConfig


STATE_PATH = Path("_system/state/workspace-sources.json")


@dataclass(frozen=True)
class WorkspaceFile:
    path: str
    relative_path: str
    modified_at: str
    size: int
    extension: str
    sha256: str
    reason: str
    recommended_source_type: str
    already_sourced: bool
    source_id: str | None = None
    source_path: str | None = None

    def to_agent_dict(self) -> dict[str, object]:
        return {
            "path": self.path,
            "relativePath": self.relative_path,
            "modifiedAt": self.modified_at,
            "size": self.size,
            "extension": self.extension,
            "sha256": self.sha256,
            "reason": self.reason,
            "recommendedSourceType": self.recommended_source_type,
            "alreadySourced": self.already_sourced,
            "sourceId": self.source_id,
            "sourcePath": self.source_path,
        }


def default_workspace_root(config: AgentWikiConfig, explicit_root: Path | None = None) -> Path:
    if explicit_root is not None:
        return explicit_root.resolve()
    if config.workspace_root is not None:
        return config.workspace_root.resolve()
    if config.wiki_type == "workspace":
        return config.root.resolve()
    return Path.cwd().resolve()


def wiki_root_for_workspace(workspace_root: Path, wiki_dir: str | None = None) -> Path:
    clean_wiki_dir = (wiki_dir or DEFAULT_WORKSPACE_WIKI_DIR).strip().strip("/") or DEFAULT_WORKSPACE_WIKI_DIR
    return (workspace_root / clean_wiki_dir).resolve()


def load_state(wiki_root: Path) -> dict[str, object]:
    path = wiki_root / STATE_PATH
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return {"schemaVersion": 1, "files": {}}
    if not isinstance(data, dict):
        return {"schemaVersion": 1, "files": {}}
    if not isinstance(data.get("files"), dict):
        data["files"] = {}
    return data


def write_state(wiki_root: Path, state: dict[str, object]) -> None:
    path = wiki_root / STATE_PATH
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(state, indent=2, sort_keys=True) + "\n", encoding="utf-8")


def scan_workspace(
    workspace_root: Path,
    wiki_root: Path,
    scan_config: WorkspaceScanConfig,
    *,
    since: datetime | None = None,
    state: dict[str, object] | None = None,
) -> list[WorkspaceFile]:
    workspace_root = workspace_root.resolve()
    wiki_root = wiki_root.resolve()
    state = state or {"files": {}}
    files_state = state.get("files") if isinstance(state.get("files"), dict) else {}
    results: list[WorkspaceFile] = []

    for path in iter_candidate_paths(workspace_root, wiki_root, scan_config):
        try:
            stat = path.stat()
        except OSError:
            continue
        modified = datetime.fromtimestamp(stat.st_mtime, tz=timezone.utc)
        if since is not None and modified < since:
            continue
        rel_path = path.relative_to(workspace_root).as_posix()
        digest = sha256_file(path)
        previous = files_state.get(rel_path) if isinstance(files_state, dict) else None
        previous_hash = previous.get("sha256") if isinstance(previous, dict) else None
        already_sourced = bool(previous.get("sourceId")) if isinstance(previous, dict) else False
        source_id = previous.get("sourceId") if isinstance(previous, dict) else None
        source_path = previous.get("sourcePath") if isinstance(previous, dict) else None
        if previous_hash is None:
            reason = "new"
        elif previous_hash != digest:
            reason = "changed"
        else:
            reason = "unchanged"
        results.append(
            WorkspaceFile(
                path=str(path),
                relative_path=rel_path,
                modified_at=modified.isoformat(),
                size=stat.st_size,
                extension=path.suffix.lower(),
                sha256=digest,
                reason=reason,
                recommended_source_type=recommend_source_type(path),
                already_sourced=already_sourced,
                source_id=str(source_id) if source_id else None,
                source_path=str(source_path) if source_path else None,
            )
        )

    return sorted(results, key=lambda item: item.relative_path)


def update_state_from_scan(wiki_root: Path, files: Iterable[WorkspaceFile], state: dict[str, object]) -> dict[str, object]:
    files_state = state.get("files") if isinstance(state.get("files"), dict) else {}
    now = datetime.now(timezone.utc).isoformat()
    for item in files:
        previous = files_state.get(item.relative_path) if isinstance(files_state, dict) else None
        source_id = previous.get("sourceId") if isinstance(previous, dict) else None
        source_path = previous.get("sourcePath") if isinstance(previous, dict) else None
        files_state[item.relative_path] = {
            "path": item.path,
            "mtime": item.modified_at,
            "size": item.size,
            "sha256": item.sha256,
            "sourceId": source_id,
            "sourcePath": source_path,
            "lastSeenAt": now,
        }
    state["schemaVersion"] = 1
    state["lastScanAt"] = now
    state["files"] = files_state
    write_state(wiki_root, state)
    return state


def mark_sourced(
    wiki_root: Path,
    state: dict[str, object],
    *,
    relative_path: str,
    source_id: str,
    source_path: str,
) -> dict[str, object]:
    files_state = state.get("files") if isinstance(state.get("files"), dict) else {}
    previous = files_state.get(relative_path) if isinstance(files_state, dict) else None
    record = dict(previous) if isinstance(previous, dict) else {}
    record["sourceId"] = source_id
    record["sourcePath"] = source_path
    record["mappedAt"] = datetime.now(timezone.utc).isoformat()
    files_state[relative_path] = record
    state["schemaVersion"] = 1
    state["files"] = files_state
    write_state(wiki_root, state)
    return state


def iter_candidate_paths(workspace_root: Path, wiki_root: Path, scan_config: WorkspaceScanConfig) -> Iterable[Path]:
    excluded_dirs = set(scan_config.exclude_dirs)
    for path in workspace_root.rglob("*"):
        if not path.is_file():
            continue
        if is_relative_to(path, wiki_root):
            continue
        rel_parts = path.relative_to(workspace_root).parts
        if any(part in excluded_dirs for part in rel_parts[:-1]):
            continue
        if any(fnmatch(path.name, pattern) for pattern in scan_config.exclude_file_globs):
            continue
        if path.suffix.lower() not in scan_config.include_extensions:
            continue
        yield path


def is_relative_to(path: Path, parent: Path) -> bool:
    try:
        path.resolve().relative_to(parent.resolve())
    except ValueError:
        return False
    return True


def sha256_file(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def recommend_source_type(path: Path) -> str:
    suffix = path.suffix.lower()
    if suffix == ".pdf":
        return "pdf"
    if suffix in {".csv", ".json", ".yaml", ".yml"}:
        return "dataset"
    if suffix in {".md", ".markdown", ".txt", ".docx"}:
        return "document"
    return "other"


def files_to_json(files: Iterable[WorkspaceFile]) -> str:
    return json.dumps([item.to_agent_dict() for item in files], indent=2, sort_keys=True)


def files_to_text(files: Iterable[WorkspaceFile]) -> str:
    rows = list(files)
    if not rows:
        return "No workspace source candidates found."
    lines = []
    for item in rows:
        marker = "sourced" if item.already_sourced else item.reason
        lines.append(f"{marker:9} {item.relative_path} ({item.recommended_source_type}, {item.size} bytes)")
    return "\n".join(lines)

File: src/workspace.ts
import { createHash } from "node:crypto";
import { existsSync, mkdirSync, readFileSync, readdirSync, statSync, writeFileSync } from "node:fs";
import { dirname, join, relative, resolve } from "node:path";
import { DEFAULT_WORKSPACE_WIKI_DIR, AgentWikiConfig, WorkspaceScanConfig, cleanWikiDir, isObject } from "./config.js";

export const STATE_PATH = "_system/state/workspace-sources.json";

export interface WorkspaceFile {
  path: string;
  relativePath: string;
  modifiedAt: string;
  size: number;
  extension: string;
  sha256: string;
  reason: "new" | "changed" | "unchanged";
  recommendedSourceType: string;
  alreadySourced: boolean;
  sourceId: string | null;
  sourcePath: string | null;
}

export function defaultWorkspaceRoot(config: AgentWikiConfig, explicitRoot?: string): string {
  if (explicitRoot) {
    return resolve(explicitRoot);
  }
  if (config.workspaceRoot) {
    return resolve(config.workspaceRoot);
  }
  if (config.wikiType === "workspace") {
    return resolve(config.root);
  }
  return process.cwd();
}

export function wikiRootForWorkspace(workspaceRoot: string, wikiDir?: string): string {
  return resolve(workspaceRoot, cleanWikiDir(wikiDir ?? DEFAULT_WORKSPACE_WIKI_DIR));
}

export function loadState(wikiRoot: string): Record<string, unknown> {
  const path = join(wikiRoot, STATE_PATH);
  try {
    const data: unknown = JSON.parse(readFileSync(path, "utf8"));
    if (!isObject(data)) {
      return emptyState();
    }
    if (!isObject(data.files)) {
      data.files = {};
    }
    return data;
  } catch {
    return emptyState();
  }
}

export function writeState(wikiRoot: string, state: Record<string, unknown>): void {
  const path = join(wikiRoot, STATE_PATH);
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, `${JSON.stringify(state, null, 2)}\n`, "utf8");
}

export function scanWorkspace(
  workspaceRootInput: string,
  wikiRootInput: string,
  scanConfig: WorkspaceScanConfig,
  options: { since?: Date; state?: Record<string, unknown> } = {}
): WorkspaceFile[] {
  const workspaceRoot = resolve(workspaceRootInput);
  const wikiRoot = resolve(wikiRootInput);
  const filesState = isObject(options.state?.files) ? options.state.files : {};
  const results: WorkspaceFile[] = [];
  for (const path of iterCandidatePaths(workspaceRoot, wikiRoot, scanConfig)) {
    let stat;
    try {
      stat = statSync(path);
    } catch {
      continue;
    }
    const modified = new Date(stat.mtimeMs);
    if (options.since && modified < options.since) {
      continue;
    }
    const relPath = relative(workspaceRoot, path).split("\\").join("/");
    const digest = sha256File(path);
    const previous = isObject(filesState[relPath]) ? filesState[relPath] : null;
    const previousHash = typeof previous?.sha256 === "string" ? previous.sha256 : null;
    const sourceId = typeof previous?.sourceId === "string" ? previous.sourceId : null;
    const sourcePath = typeof previous?.sourcePath === "string" ? previous.sourcePath : null;
    const reason = previousHash === null ? "new" : previousHash === digest ? "unchanged" : "changed";
    results.push({
      path,
      relativePath: relPath,
      modifiedAt: modified.toISOString(),
      size: stat.size,
      extension: extensionOf(path),
      sha256: digest,
      reason,
      recommendedSourceType: recommendSourceType(path),
      alreadySourced: sourceId !== null,
      sourceId,
      sourcePath
    });
  }
  return results.sort((a, b) => a.relativePath.localeCompare(b.relativePath));
}

export function updateStateFromScan(wikiRoot: string, files: WorkspaceFile[], state: Record<string, unknown>): Record<string, unknown> {
  const filesState = isObject(state.files) ? state.files : {};
  const now = new Date().toISOString();
  for (const item of files) {
    const previous = (isObject(filesState[item.relativePath]) ? filesState[item.relativePath] : {}) as Record<string, unknown>;
    filesState[item.relativePath] = {
      path: item.path,
      mtime: item.modifiedAt,
      size: item.size,
      sha256: item.sha256,
      sourceId: typeof previous.sourceId === "string" ? previous.sourceId : null,
      sourcePath: typeof previous.sourcePath === "string" ? previous.sourcePath : null,
      lastSeenAt: now
    };
  }
  state.schemaVersion = 1;
  state.lastScanAt = now;
  state.files = filesState;
  writeState(wikiRoot, state);
  return state;
}

export function markSourced(
  wikiRoot: string,
  state: Record<string, unknown>,
  options: { relativePath: string; sourceId: string; sourcePath: string }
): Record<string, unknown> {
  const filesState = isObject(state.files) ? state.files : {};
  const previous = (isObject(filesState[options.relativePath]) ? filesState[options.relativePath] : {}) as Record<string, unknown>;
  filesState[options.relativePath] = {
    ...previous,
    sourceId: options.sourceId,
    sourcePath: options.sourcePath,
    mappedAt: new Date().toISOString()
  };
  state.schemaVersion = 1;
  state.files = filesState;
  writeState(wikiRoot, state);
  return state;
}

export function filesToJson(files: WorkspaceFile[]): string {
  return JSON.stringify(files, null, 2);
}

export function filesToText(files: WorkspaceFile[]): string {
  if (files.length === 0) {
    return "No workspace source candidates found.";
  }
  return files
    .map((item) => {
      const marker = item.alreadySourced ? "sourced" : item.reason;
      return `${marker.padEnd(9)} ${item.relativePath} (${item.recommendedSourceType}, ${item.size} bytes)`;
    })
    .join("\n");
}

function emptyState(): Record<string, unknown> {
  return { schemaVersion: 1, files: {} };
}

function iterCandidatePaths(workspaceRoot: string, wikiRoot: string, scanConfig: WorkspaceScanConfig): string[] {
  const results: string[] = [];
  const excludedDirs = new Set(scanConfig.excludeDirs);
  function walk(dir: string): void {
    let entries;
    try {
      entries = readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      const path = join(dir, entry.name);
      if (entry.isDirectory()) {
        const relParts = relative(workspaceRoot, path).split(/[\\/]/);
        if (isRelativeTo(path, wikiRoot) || relParts.some((part) => excludedDirs.has(part))) {
          continue;
        }
        walk(path);
      } else if (entry.isFile()) {
        const relParts = relative(workspaceRoot, path).split(/[\\/]/);
        if (isRelativeTo(path, wikiRoot) || relParts.slice(0, -1).some((part) => excludedDirs.has(part))) {
          continue;
        }
        if (scanConfig.excludeFileGlobs.some((pattern) => globMatch(entry.name, pattern))) {
          continue;
        }
        if (!scanConfig.includeExtensions.includes(extensionOf(path))) {
          continue;
        }
        results.push(resolve(path));
      }
    }
  }
  walk(workspaceRoot);
  return results;
}

function isRelativeTo(path: string, parent: string): boolean {
  const rel = relative(resolve(parent), resolve(path));
  return rel === "" || (!rel.startsWith("..") && !rel.startsWith("/") && !/^[A-Za-z]:/.test(rel));
}

function sha256File(path: string): string {
  return createHash("sha256").update(readFileSync(path)).digest("hex");
}

function recommendSourceType(path: string): string {
  const suffix = extensionOf(path);
  if (suffix === ".pdf") return "pdf";
  if ([".csv", ".json", ".yaml", ".yml"].includes(suffix)) return "dataset";
  if ([".md", ".markdown", ".txt", ".docx"].includes(suffix)) return "document";
  return "other";
}

function extensionOf(path: string): string {
  const match = /(\.[^./\\]+)$/.exec(path);
  return match ? match[1].toLowerCase() : "";
}

function globMatch(name: string, pattern: string): boolean {
  if (pattern.startsWith("*")) {
    return name.endsWith(pattern.slice(1));
  }
  return name === pattern;
}
